File: C#/8.cs
public class Solution {
    public int MyAtoi(string s) {
        int index = 0;
        int sign = 1;
        int result = 0;
        
        // ignore leading whitespace
        while (index < s.Length && s[index] == ' ') {
            index++;
        }
        
        // check for sign
        if (index < s.Length && (s[index] == '+' || s[index] == '-')) {
            sign = (s[index] == '-') ? -1 : 1;
            index++;
        }
        
        // read in digits
        while (index < s.Length && char.IsDigit(s[index])) {
            int digit = s[index] - '0';
            // check for overflow
            if (result > int.MaxValue / 10 || (result == int.MaxValue / 10 && digit > 7)) {
                return (sign == 1) ? int.MaxValue : int.MinValue;
            }
            result = result * 10 + digit;
            index++;
        }
        
        return result * sign;
    }
}

File: python/8.py
class Solution:
    def myAtoi(self, s: str) -> int:
        #remove leading whitespace
        s = s.lstrip()
        #check for sign
        if s and (s[0] == '-' or s[0] == '+'):
            sign = s[0]
            s = s[1:]
        else:
            sign = ''
        #parse digits
        num_str = ''
        for c in s:
            if c.isdigit():
                num_str += c
            else:
                break
        #convert to int and apply sign
        if num_str:
            num = int(sign + num_str)
        else:
            num = 0
        #clamp to range
        INT_MIN = -2**31
        INT_MAX = 2**31 - 1
        if num < INT_MIN:
            return INT_MIN
        elif num > INT_MAX:
            return INT_MAX
        else:
            return num